public/src/components/ToastManager.jsx

```diff
@@ -0,0 +1,198 @@
+import { useState, useRef } from 'react';
+import * as React from 'react';
+import { PropTypes } from 'prop-types';
+import styled from 'styled-components';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import {
+  faCheckCircle, faExclamationTriangle, faTimes,
+} from '@fortawesome/free-solid-svg-icons';
+
+const getCSSForFlavor = (flavor) => {
+  const flavors = {
+    success: `
+      background-color: #386C0B;
+    `,
+    error: `
+      background-color: #EF6F6C;
+    `,
+  };
+  return flavors[flavor];
+};
+
+export const ToastContext = React.createContext();
+const ToastRack = styled.div`
+  z-index: 1;
+  position: fixed;
+  right: 0;
+  width: 250px;
+  bottom: 0;
+  padding: 1em;
+`;
+
+const StyledToast = styled.div`
+  @keyframes styledtoast {
+    0% {
+      transform: translateY(20px)
+    }
+    100% {
+      transform: translateY(0px)
+    }
+  }
+  @keyframes cancelled {
+  0% {
+    transform: translateX(0px);
+  }
+  90% {
+    transform: translateX(0px);
+  }
+  100% {
+    transform: translateX(300px);
+  }
+}
+  @keyframes immediatecancel {
+  0% {
+    transform: translateX(0px);
+  }
+  100% {
+    transform: translateX(300px);
+  }
+}
+
+  display: flex;
+  border: 1px solid #54457F;
+  width: 250px;
+  cursor: pointer;
+  animation: styledtoast .3s ease-out, cancelled 4.5s ease-out .3s;
+  animation-fill-mode: forwards;  
+  border-radius: .25em;
+  background-color: white;
+  box-shadow: 0px 4px 3px 3px rgba(0,0,0, .1);
+  & + & {
+    margin-top: 1em;
+  }
+
+  & .toast__alert {
+    margin-left: -1px;
+    border-radius: .25em 0 0 .25em;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    width: 25px;
+    & path {
+      fill: white;
+    }
+    padding: .25em .5em;
+    ${(props) => getCSSForFlavor(props.flavor)}
+  }
+  & .toast__content {
+    padding: .5em 1em;
+    flex: 1;
+  }
+  &.toast--cancelled {
+    animation: immediatecancel .35s ease-out;
+  }
+  & .toast__cancel {
+    display: flex;
+    padding: .25em .5em;
+    & path {
+      fill: #54457F;
+    }
+    align-items: center;
+  }
+`;
+
+export const ToastAlert = ({ children, flavor, id }) => {
+  const { manuallyCancel, flavors } = React.useContext(ToastContext);
+  const internalFlavors = {
```

Comment:
Doesn't need to be defined inside of the component

Reply:
Philosophical question though, should it? I figured it should because both of those are important for specifically that component.

Reply:
No. You're defining a new object reference needlessly on every re-render of a component. If the values are not dependent on anything pertaining to props/state/context, then there is no reason to define the constant literal inside of the component.

That would be like defining every single component definition inside of the parent component if it's only ever rendered by that component (they are just consts, after all)

Reply:
Ah capiche.

Reply:
Wait but when i try and move that stuff out, I get complaints because useContext shouldn't be used at top level, has to be used inside a component.

Reply:
Cuz of a refactor that i did, you'll see. One sec.

Reply:
Yeah, you can't use `useContext` outside of a component (true of any hooks).